### PR TITLE
Fix TestServiceAlloc flakes

### LINKED ISSUE #37040: TestServiceAlloc {master}
https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-test-go/95/

Failed: TestServiceAlloc {master}

```
master_test.go:484: unexpected creation success. The following IPs exist: []string{"192.168.0.1", "192.168.0.1", "192.168.0.6", "192.168.0.5", "192.168.0.3", "192.168.0.4", "192.168.0.2"}. It should only be possible to allocate 2 IP addresses in this cluster.
	
	&api.ServiceList{TypeMeta:unversioned.TypeMeta{Kind:"", APIVersion:""}, ListMeta:unversioned.ListMeta{SelfLink:"/api/v1/services", ResourceVersion:"1010"}, Items:[]api.Service{api.Service{TypeMeta:unversioned.TypeMeta{Kind:"", APIVersion:""}, ObjectMeta:api.ObjectMeta{Name:"kubernetes", GenerateName:"", Namespace:"default", SelfLink:"/api/v1/namespaces/default/services/kubernetes", UID:"f3602349-ad19-11e6-a96b-0242ac110002", ResourceVersion:"986", Generation:0, CreationTimestamp:unversioned.Time{Time:time.Time{sec:63615020565, nsec:0, loc:(*time.Location)(0x4786a60)}}, DeletionTimestamp:(*unversioned.Time)(nil), DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string{"component":"apiserver", "provider":"kubernetes"}, Annotations:map[string]string(nil), OwnerReferences:[]api.OwnerReference(nil), Finalizers:[]string(nil), ClusterName:""}, Spec:api.ServiceSpec{Type:"ClusterIP", Ports:[]api.ServicePort{api.ServicePort{Name:"https", Protocol:"TCP", Port:443, TargetPort:intstr.IntOrString{Type:0, IntVal:443, StrVal:""}, NodePort:0}}, Selector:map[string]string(nil), ClusterIP:"192.168.0.1", ExternalName:"", ExternalIPs:[]string(nil), LoadBalancerIP:"", SessionAffinity:"ClientIP", LoadBalancerSourceRanges:[]string(nil)}, Status:api.ServiceStatus{LoadBalancer:api.LoadBalancerStatus{Ingress:[]api.LoadBalancerIngress(nil)}}}, api.Service{TypeMeta:unversioned.TypeMeta{Kind:"", APIVersion:""}, ObjectMeta:api.ObjectMeta{Name:"svc-0", GenerateName:"", Namespace:"default", SelfLink:"/api/v1/namespaces/default/services/svc-0", UID:"f392cd35-ad19-11e6-a96b-0242ac110002", ResourceVersion:"1000", Generation:0, CreationTimestamp:unversioned.Time{Time:time.Time{sec:63615020565, nsec:0, loc:(*time.Location)(0x4786a60)}}, DeletionTimestamp:(*unversioned.Time)(nil), DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string(nil), Annotations:map[string]string(nil), OwnerReferences:[]api.OwnerReference(nil), Finalizers:[]string(nil), ClusterName:""}, Spec:api.ServiceSpec{Type:"ClusterIP", Ports:[]api.ServicePort{api.ServicePort{Name:"", Protocol:"TCP", Port:80, TargetPort:intstr.IntOrString{Type:0, IntVal:80, StrVal:""}, NodePort:0}}, Selector:map[string]string(nil), ClusterIP:"192.168.0.1", ExternalName:"", ExternalIPs:[]string(nil), LoadBalancerIP:"", SessionAffinity:"None", LoadBalancerSourceRanges:[]string(nil)}, Status:api.ServiceStatus{LoadBalancer:api.LoadBalancerStatus{Ingress:[]api.LoadBalancerIngress(nil)}}}, api.Service{TypeMeta:unversioned.TypeMeta{Kind:"", APIVersion:""}, ObjectMeta:api.ObjectMeta{Name:"svc-1", GenerateName:"", Namespace:"default", SelfLink:"/api/v1/namespaces/default/services/svc-1", UID:"f393d738-ad19-11e6-a96b-0242ac110002", ResourceVersion:"1002", Generation:0, CreationTimestamp:unversioned.Time{Time:time.Time{sec:63615020565, nsec:0, loc:(*time.Location)(0x4786a60)}}, DeletionTimestamp:(*unversioned.Time)(nil), DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string(nil), Annotations:map[string]string(nil), OwnerReferences:[]api.OwnerReference(nil), Finalizers:[]string(nil), ClusterName:""}, Spec:api.ServiceSpec{Type:"ClusterIP", Ports:[]api.ServicePort{api.ServicePort{Name:"", Protocol:"TCP", Port:80, TargetPort:intstr.IntOrString{Type:0, IntVal:80, StrVal:""}, NodePort:0}}, Selector:map[string]string(nil), ClusterIP:"192.168.0.6", ExternalName:"", ExternalIPs:[]string(nil), LoadBalancerIP:"", SessionAffinity:"None", LoadBalancerSourceRanges:[]string(nil)}, Status:api.ServiceStatus{LoadBalancer:api.LoadBalancerStatus{Ingress:[]api.LoadBalancerIngress(nil)}}}, api.Service{TypeMeta:unversioned.TypeMeta{Kind:"", APIVersion:""}, ObjectMeta:api.ObjectMeta{Name:"svc-2", GenerateName:"", Namespace:"default", SelfLink:"/api/v1/namespaces/default/services/svc-2", UID:"f39481fe-ad19-11e6-a96b-0242ac110002", ResourceVersion:"1004", Generation:0, CreationTimestamp:unversioned.Time{Time:time.Time{sec:63615020565, nsec:0, loc:(*time.Location)(0x4786a60)}}, DeletionTimestamp:(*unversioned.Time)(nil), DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string(nil), Annotations:map[strin
```

Previous issues for this test: #31850 #31880 #32028

### PR DESCRIPTION
**Pull request by wojtek-t**
Fix #37040

```release-note
Fix TestServiceAlloc flakes
```

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/37487)

### CODE REVIEW

**Comment by smarterclayton** [comment]
Lgtm

**Comment by wojtek-t** [comment]
@k8s-bot unit test this, issue: #IGNORE (Waiting for status to be reported)

File: pkg/registry/generic/registry/store.go [hunk 1]
```diff
@@ -203,7 +203,8 @@ func (e *Store) List(ctx api.Context, options *api.ListOptions) (runtime.Object,
 // ListPredicate returns a list of all the items matching m.
 func (e *Store) ListPredicate(ctx api.Context, p storage.SelectionPredicate, options *api.ListOptions) (runtime.Object, error) {
 	if options == nil {
-		options = &api.ListOptions{ResourceVersion: "0"}
+		// By default we should serve the request from etcd.
+		options = &api.ListOptions{ResourceVersion: ""}
```
Comment (deads2k) [inline pkg/registry/generic/registry/store.go]: @liggitt @ncdc

**Comment by wojtek-t** [comment]
@deads2k @smarterclayton - if you're fine with this, can one of you apply label?

File: pkg/registry/core/service/ipallocator/controller/repair.go [hunk 2]
```diff
@@ -105,7 +105,7 @@ func (c *Repair) runOnce() error {
 	// the service collection. The caching layer keeps per-collection RVs,
 	// and this is proper, since in theory the collections could be hosted
 	// in separate etcd (or even non-etcd) instances.
-	list, err := c.registry.ListServices(ctx, nil)
```
Comment (liggitt) [inline pkg/registry/core/service/ipallocator/controller/repair.go]: With the fix in the generic store, why is this needed?

File: pkg/registry/generic/registry/store.go (hunk 1 above)
Comment (liggitt) [inline pkg/registry/generic/registry/store.go]: Does this need backporting?

**Comment by deads2k** [comment]
> @deads2k @smarterclayton - if you're fine with this, can one of you apply label?

@liggitt is more familiar with this area than I am.  I'll defer to him.

File: pkg/registry/core/service/ipallocator/controller/repair.go (hunk 2 above)
Comment (wojtek-t) [inline pkg/registry/core/service/ipallocator/controller/repair.go]: We just should avoid passing nil - it seems there are just few such places in the code and we should forbid it.

**Comment by wojtek-t** [comment]
> Does this need backporting?

Good question - maybe? But it's not that crucial, because if you are http sending a request to apiserver, the the "List handler" will always create ListOptions and for that path it's never nil.
So it's only a matter of some calls coming from apiserver internals.

**Comment by liggitt** [comment]
> So it's only a matter of some calls coming from apiserver internals.

If one of those internal paths was the IP allocator, that seems important to backport

**Comment by wojtek-t** [comment]
>If one of those internal paths was the IP allocator, that seems important to backport

I wasn't trying to say not to backport it - I'm fine with backporting.

**Comment by wojtek-t** [comment]
But first we probably need to have in in head :)

**Comment by liggitt** [comment]
LGTM

**Comment by liggitt** [comment]
I would pick it to 1.5 and 1.4

**Comment by wojtek-t** [comment]
@liggitt - definitely for 1.5. But I'm fine with cherrypicking it to 1.4 too - I will prepare cherrypicks once this is merged.

File: pkg/registry/generic/registry/store.go (hunk 1 above)
Comment (smarterclayton) [inline pkg/registry/generic/registry/store.go]: I think so.

**Comment by dims** [comment]
@k8s-bot test this

**Comment by wojtek-t** [comment]
@dims - why did you trigger test on all PRs? This is why we ended in a situation where a bunch of tests on different PRs are failing with "error in queue" or "waiting for status" or sth like that.

@k8s-bot test this, issue: #IGNORE

**Comment by dims** [comment]
@wojtek-t : just 1.5 milestone with lgtm(s), trying to get the submit queue to pick it up.

**Comment by wojtek-t** [comment]
@dims - you do NOT need to trigger tests to merge bot to pick them up. If retest will be needed, merge bot will do it itself. What you did actually pushed a bunch of PRs OUT of queue, because after your trigger a bunch of them stuck in "error in queue" or sth like that state. So please do NOT do it again.

**Comment by wojtek-t** [comment]
@k8s-bot test this, issue: #IGNORE (build started)

**Comment by wojtek-t** [comment]
@k8s-bot e2e test this, issue: #IGNORE (error in queue)

**Comment by wojtek-t** [comment]
@k8s-bot etcd3 test this, issue: #33380

**Comment by dims** [comment]
@wojtek-t ack. will not repeat that. apologies.

**Pull request merged**